### PR TITLE
Detect typo's and suggest correct naming

### LINKED ISSUE #893: Detect typo in metadata
Currently, the metadata linter warns you on unknown meta. However, this can be confusing if the linter is outdated. Maybe we can try to calculate string distance (or something like that) and only warn users about typos.

Ref: https://github.com/StylishThemes/Overlay-Scrollbars/pull/10#issuecomment-606559726

### PR DESCRIPTION
**Pull request by Gusted**
Using the [Levensthein Distance algorithm](https://en.wikipedia.org/wiki/Levenshtein_distance) we can detect if 2 strings could be made with an x amount of edits. The implementation I made is a variation of the the 2 matrix-rows implementation so it's optimal to check if we can "bail out" if the maxEdits is too high.

It should resolve https://github.com/openstyles/stylus/issues/893
Even though that issue mentions to only error on typo's I'm not sure if that's correct and should be the behavior, but It would take little edit to have such behavior.

Also I've for now opted for 2 edits as if someone makes a typo I don't expect that typo to be more out of placed than 2 edits but could be very easily set to another number if needed.

Regards,
Gusted

### CODE REVIEW

**Comment by Gusted** [comment]
Seems like tests are failing on files I didn't touch. I might have magic hands...

**Comment by Gusted** [comment]
That should fix the failing tests for now.

**Review by eight04 — CHANGES_REQUESTED** [review]
👍

----
I'd like to discuss an API issue, anything relates to the following issue should probably be implemented in another PR: What is the expected behavior of the following code?
```js
parse(`
  /* ==UserStyle==
  @namu This is my style's name.
  ==/UserStyle== */
`, {unknownKey: 'assign' /* or 'ignore' */, mandatoryKeys: []})
```
Currently, you can only get the typo suggestion with `unknownKey: 'throw'`. IMO, users should be able to catch the typo warning *and* ignore/assign/throw.

Maybe we can

* Make the parser be able to emit warnings and implement an `onwarn` handler?
* Allow users to configure a function to parse unknown keys? This will make it possible to autofix key name. But do we need autofix?
* Other ideas?

File: lib/levensthein-distance.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/**
+ * Gives you a array with filled with 1 to amount.
+ * @param {number} amount
+ * @returns {number[]}
+ */
+function range(amount) {
+  const range = Array.from({amount});
```
[the remark this answers is not in the document]
Comment (Gusted) [reply lib/levensthein-distance.js:7]: 🔔 Ding ding.

**Comment by Gusted** [comment]
The best implementation seems in my eyes to be independent of the `unknownKey` option as we are suggesting(if any) the correct key and could be implemented with an `onWarn` or even a new option like `autoFix` or `error-and-show-suggestion` and add it inside the thrown error. It's more what's practical for the end user. If the end-user sees such error and the error message shows an suggestive change it seems good in my eyes, even better would be if the user saves the document it fixes the metadata(which on it's own could be unexpected behavior).

**Comment by Gusted** [comment]
A more discrete answer on ^ 

having an onWarn callback. While providing/default value the auto-fix with Levensthein distance detection.

File: tests/levensthein-distance.test.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const test = require('ava');
+const {LevenshteinDistanceWithMax} = require('../lib/levensthein-distance');
+
+test("shouldn't match", t => {
+  t.is(LevenshteinDistanceWithMax('water', 'atect', 2), false);
+  t.is(LevenshteinDistanceWithMax('water', 'christmas', 3), false);
+  t.is(LevenshteinDistanceWithMax('water', 'water1', 0), false);
+  t.is(LevenshteinDistanceWithMax('thea', 'ythee', 1), false);
+  t.is(LevenshteinDistanceWithMax('12345', '567', 4), false);
+  t.is(LevenshteinDistanceWithMax('var', 'abc', 3), false);
```
Comment (eight04) [inline tests/levensthein-distance.test.js:10]: This should be true. You only have to do 3 substitutions from `var` to `abc`.

File: README.md [hunk 3]
```diff
@@ -243,7 +243,7 @@ Use this class to initiate a parse error.
 * `code` - error code.
 * `message` - error message.
 * `index` - the string index where the error occurs.
-* `args` - an array of values that is used to compose the error message. This allows other clients to generate i18n error message.
+* `args` - an array of values that is used to compose the error message. This allows other clients to generate i18n error message. If the error code is unknownXXX the last value of this array would be a suggestive change in which the correct value is set, if it couldn't be find it will be `undefined` and couldn't find an suggestive change.
```
Comment (eight04) [inline README.md:246]:
```diff
diff --git a/README.md b/README.md
index 659232e..ef339d8 100644
--- a/README.md
+++ b/README.md
@@ -243,7 +243,7 @@ Use this class to initiate a parse error.
 * `code` - error code.
 * `message` - error message.
 * `index` - the string index where the error occurs.
-* `args` - an array of values that is used to compose the error message. This allows other clients to generate i18n error message. If the error code is unknownXXX the last value of this array would be a suggestive change in which the correct value is set, if it couldn't be find it will be `undefined` and couldn't find an suggestive change.
+* `args` - an array of values that is used to compose the error message. This allows other clients to generate i18n error message.
 
 A table of errors thrown by the parser:
 
@@ -275,7 +275,7 @@ A table of errors thrown by the parser:
 |`missingMandatory`|A list of missing keys|This error doesn't have `err.index`.|
 |`missingValue`||Expect a non-whitespace value.|
 |`unknownJSONLiteral`|Literal value|JSON has only 3 literals: `true`, `false`, and `null`.|
-|`unknownMeta`|Key of unknown metadata|Unknown `@metadata`.|
+|`unknownMeta`|`[META_KEY, SUGGESTED_META_KEY]`|Unknown `@metadata`. It may suggest the correct metadata name if there is a typo. `SUGGESTED_META_KEY` can be null|
 |`unknownVarType`|`[META_KEY, VARIABLE_TYPE]`|Unknown variable type. `META_KEY` could be `var` or `advanced`.|
 
 ### util

```

File: lib/levensthein-distance.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/**
+ * Gives you a array with filled with 0...amount.
+ * @param {number} amount
+ * @returns {number[]}
+ */
+function range(amount) {
+  const range = Array.from({length: amount});
```
Comment (eight04) [inline lib/levensthein-distance.js:7]: Just use `Array(amount)`. There is no advantage to use `Array.from` here.

1. You have to create a new object.
2. `Array.from` has to visit each index and copy the value (undefined) from the object to the array.

File: lib/parse.js [hunk 5]
```diff
@@ -363,6 +364,8 @@ function isMultipleOf(value, step) {
   return Math.abs(n - nInt) < Math.pow(10, (`${nInt}`.length - 16));
 }
 
+const MAX_EDIT = 3;
```
Comment (eight04) [inline lib/parse.js:367]: I think 1 is good, 2 should be fine, 3 is too much to be considered as a typo. @tophf

File: lib/parse.js [hunk 6]
```diff
@@ -530,10 +534,14 @@ function createParser({
...
+        // TODO: Suggest the item with the smallest distance or even multiple results?
+        // Implementation note: swtich to Levenshtein automaton variation.
+        const maybeSuggestion = keysOfParser.find(metaKey => LevenshteinDistanceWithMax(metaKey, state.key, MAX_EDIT));
+
         // throw
         throw new ParseError({
           code: 'unknownMeta',
-          args: [state.key],
+          args: [state.key, maybeSuggestion],
           message: `Unknown metadata: @${state.key}`,
```
Comment (eight04) [inline lib/parse.js:545]: Don't revert the change to the message. We put the variable into `vars` if it is used to compose `message`.

File: tests/levensthein-distance.test.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+});
+
+test('should match', t => {
+  t.is(LevenshteinDistanceWithMax('advanced', 'advance', 3), true);
+  t.is(LevenshteinDistanceWithMax('water', 'water', 0), true);
+  t.is(LevenshteinDistanceWithMax('wayer', 'water', 1), true);
```
Comment (eight04) [inline tests/levensthein-distance.test.js:16]: Add a new case:
```js
  t.is(LevenshteinDistanceWithMax('wayter', 'water', 1), true);
```

**Review by eight04 — CHANGES_REQUESTED** [review]

File: lib/parse.js (hunk 5 above)
Comment (tophf) [reply lib/parse.js:367]: Why is `3` too much? It's certainly possible to make 3 typos so ideally I would try to calculate the distance of the key on the keyboard, which unfortunately isn't universally usable as notebooks tend to have unique layouts.

Comment (eight04) [reply lib/parse.js:367]: Note that this number applies to a single word. For example, the edit distance from `var` to `name` is also 3 (v -> n, r -> m, insert e).

Comment (tophf) [reply lib/parse.js:367]: Can it be dynamic? Like `word.length / 2`?

File: tests/levensthein-distance.test.js (hunk 2 above)
Comment (Gusted) [reply tests/levensthein-distance.test.js:10]: That's correct, but knowinlgy if you make an typo like `abc` I would be surprised you meant `var`. But we can make it `true`

File: lib/levensthein-distance.js (hunk 4 above)
Comment (Gusted) [reply lib/levensthein-distance.js:7]: The `unicorn/no-new-array` rule is enabled thus not allowing having `new Array(amount)`

File: lib/parse.js (hunk 5 above)
Comment (Gusted) [reply lib/parse.js:367]: Dynamically is possible, but will need test-cases for it's own. Because you would be surprised how closely certain words are in terms of `Distance cost`.

Comment (Gusted) [reply lib/parse.js:367]: `log2` seems like a good average for the max amount of `edits`

File: lib/levensthein-distance.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  const lenDiff = Math.abs(lenOne - lenTwo);
+  // Are the difference between 2 lengths greater than
+  // maxEdit, we know to bail out early on.
+  if (lenDiff > maxEdit) {
+    return false;
+  }
+
+  let prevRowDistance = range(lenOne + 1);
+  let currentRowDistance = [];
```
Comment (eight04) [inline lib/levensthein-distance.js:35]: Let's also give the current row a fixed length: `Array(lenOne + 1)`

File: lib/levensthein-distance.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  let vtemp = [];
```
Comment (eight04) [inline lib/levensthein-distance.js:36]: There is no need to define the temporary variable at the top scope.

File: lib/levensthein-distance.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+      const addCost = prevRowDistance[j] + 1;
+      const delCost = currentRowDistance[j - 1] + 1;
+      const substitionCost = prevRowDistance[j - 1] + editCost;
+
+      currentRowDistance[j] = Math.min(addCost, delCost, substitionCost);
+    }
+
+    // Swap the vectors
+    vtemp = currentRowDistance;
```
Comment (eight04) [inline lib/levensthein-distance.js:51]: `let vtemp = currentRowDistance;`

**Review by eight04 — CHANGES_REQUESTED** [review]
👍

File: lib/levensthein-distance.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  // maxEdit, we know to bail out early on.
+  if (lenDiff > maxEdit) {
+    return false;
+  }
+
+  let prevRowDistance = range(lenOne + 1);
+  let currentRowDistance = Array(lenOne + 1);
+  for (let i = 1; i <= lenTwo; i++) {
+    // Calculate the current row distances from the previous row.
+    currentRowDistance[0] = i;
```
Comment (eight04) [inline lib/levensthein-distance.js:38]: A small improvement: since the distance function is a non-decreasing function, we can bail out once the minimum distance in the current row have passed `maxEdit`.
```js
let minDistance = currentRowDistance[0] = i;
for (...) {
  ...
  currentRowDistance[j] = ...
  if (currentRowDistance[j] < minDistance) {
    minDistance = currentRowDistance[j];
  }
}
if (minDistance > maxEdit) {
  return false;
}
```

**Review by eight04 — CHANGES_REQUESTED** [review]
👍 This should be the last iteration.

File: lib/levensthein-distance.js (hunk 11 above)
Comment (Gusted) [reply lib/levensthein-distance.js:38]: Seems good to me. And still passes the test 👍🏽.

I remember I had such thing in the first commit but due to switching implementation it got lost, 😄.

BTW, you might want to revisit the eslint rules, because `let minDistance = currentRowDistance[0] = i;` will conflict with the `multi-assign` rule.
